pyomo/core/base/units_container.py

```diff
@@ -0,0 +1,1335 @@
+            Floating point tolerance used when deciding if units are equivalent
+            or not. (It can happen that units
+
+        Notes
+        -----
+        This class inherits from the :class:`StreamBasedExpressionVisitor` to implement
+        a walker that returns the pyomo units and pint units corresponding to an
+        expression.
+
+        There are class attributes (dicts) that map the expression node type to the
+        particular method that should be called to return the units of the node based
+        on the units of its child arguments. This map is used in exitNode.
+        """
+        super(_UnitExtractionVisitor, self).__init__()
+        self._pyomo_units_container = pyomo_units_container
+        self._pint_registry = self._pyomo_units_container._pint_registry
+        self._units_equivalence_tolerance = units_equivalence_tolerance
+
+    def _pint_unit_equivalent_to_dimensionless(self, pint_unit):
+        """
+        Check if a pint unit is equivalent to 'dimensionless' (this is true if it
+        is either None or 'dimensionless' (or even radians)
+
+        Parameters
+        ----------
+        pint_unit : pint unit
+           The pint unit you want to check
+
+        Returns
+        -------
+        : bool
+           Returns True if pint_unit is equivalent to dimensionless units.
+
+        """
+        if pint_unit is None:
+            return True
+        return self._pint_units_equivalent(pint_unit, self._pint_registry.dimensionless)
+
+    def _pint_units_equivalent(self, lhs, rhs):
+        """
+        Check if two pint units are equivalent
+
+        Parameters
+        ----------
+        lhs : pint unit
+            first pint unit to compare
+        rhs : pint unit
+            second pint unit to compare
+
+        Returns
+        -------
+        : bool
+           True if they are equivalent, and False otherwise
+        """
+        if lhs == rhs:
+            # units are the same objects (or both None)
+            return True
+        elif lhs is None:
+            # lhs is None, but rhs is not
+            # check if rhs is equivalent to dimensionless (e.g. dimensionless or radians)
+            return self._pint_unit_equivalent_to_dimensionless(rhs)
+        elif rhs is None:
+            # rhs is None, but lhs is not
+            # check if lhs is equivalent to dimensionless (e.g. dimensionless or radians)
+            return self._pint_unit_equivalent_to_dimensionless(lhs)
+
+        # Units are not the same objects, and they are both not None
+        # Now, use pint mechanisms to check by converting to Quantity objects
+        lhsq = (1.0 * lhs).to_base_units()
+        rhsq = (1.0 * rhs).to_base_units()
+
+        if lhsq.dimensionality == rhsq.dimensionality and \
+               abs(lhsq.magnitude/rhsq.magnitude - 1.0) < self._units_equivalence_tolerance:
+            return True
+
+        return False
+
+    def _get_units_relational(self, node, list_of_unit_tuples):
```

Comment:
This is an exact copy of `_get_unit_for_equivalent_children()`.  DRY: consider removing this method and only using that one.

Reply:
done.